Document pubsub without cloudevent. (#1499)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/pubsub/pubsub-raw.md

---
type: docs
title: "Pub/Sub without CloudEvents"
linkTitle: "Pub/Sub without CloudEvents"

> weight: 6000

Reviewer: ```suggestion
weight: 7000
```

description: "Use Pub/Sub without CloudEvents." 
---

## Introduction

Dapr uses CloudEvents to provide additional context to the event payload, enabling features like:
* Tracing
* Deduplication by message Id
* Content-type for proper deserialization of event's data

For more information about CloudEvent, please check the [official specification](https://github.com/cloudevents/spec).

> When adopting Dapr to an application, it might happen that other publisher or subscriber apps do not use Dapr and still require to communicate via Pub/Sub. For this reason, since version 1.2, Dapr enables apps to publish and subscribe to events that are not wrapped in a CloudEvent.

Reviewer: ```suggestion
When adding Dapr to your application, some services may still need to communicate via raw pub/sub messages not encapsulated in CloudEvents. This may be for compatibility reasons, or because some apps are not using Dapr yet. Dapr enables apps to publish and subscribe to raw events that are not wrapped in a CloudEvent.
```

> > Not using CloudEvent automatically disables support for tracing, deduplication per message Id, content-type metadata and any other feature built on top of the CloudEvent schema.

Reviewer: ```suggestion
{{% alert title="Warning" color="warning" %}}
Not using CloudEvents disables support for tracing, event deduplication per messageId, content-type metadata, and any other features built on top of the CloudEvent schema.
{{% /alert %}}
```

> ## Publish example

Reviewer: ```suggestion
## Publishing raw messages
```

> To disable CloudEvent wrapping, set the `rawPayload` metadata as part of the publishing request. This will enable subscribers to receive these messages without having to parsing the CloudEvent schema.

Reviewer: ```suggestion
To disable CloudEvent wrapping, set the `rawPayload` metadata to `true` as part of the publishing request. This will allow subscribers to receive these messages without having to parse the CloudEvent schema.
```

{{< tabs curl "Python SDK" "PHP SDK">}}

{{% codetab %}}
```bash
curl -X "POST" http://localhost:3500/v1.0/publish/pubsub/TOPIC_A?metadata.rawPayload=true -H "Content-Type: application/json" -d '{"order-number": "345"}'
```
{{% /codetab %}}

{{% codetab %}}
```python
from dapr.clients import DaprClient

with DaprClient() as d:
    req_data = {
        'order-number': '345'
    }
    # Create a typed message with content type and body
    resp = d.publish_event(
        pubsub_name='pubsub',
        topic='TOPIC_A',
        data=json.dumps(req_data),
        metadata=(
                     ('rawPayload', 'true')
                 )
    )
    # Print the request
    print(req_data, flush=True)
```
{{% /codetab %}}

{{% codetab %}}

```php
<?php

require_once __DIR__.'/vendor/autoload.php';

$app = \Dapr\App::create();
$app->run(function(\DI\FactoryInterface $factory) {
    $publisher = $factory->make(\Dapr\PubSub\Publish::class, ['pubsub' => 'pubsub']);
    $publisher->topic('TOPIC_A')->publish('data', ['rawPayload' => 'true']);
});
```

{{% /codetab %}}

{{< /tabs >}}

> # Subscribe example

Reviewer: ```suggestion
## Subscribing to raw events
```

> When subscribing, add the additional metadata entry for `rawPayload` and Dapr sidecar will automatically wrap the payloads into a CloudEvent that is compatible with current Dapr SDKs.

Reviewer: ```suggestion
When subscribing programmatically, add the additional metadata entry for `rawPayload` so the Dapr sidecar will automatically wrap the payloads into a CloudEvent that is compatible with current Dapr SDKs.
```

> When subscribing, add the additional metadata entry for `rawPayload` and Dapr sidecar will automatically wrap the payloads into a CloudEvent that is compatible with current Dapr SDKs.

Reviewer: Is the behavior here that setting rawPayload=true will cause the sidecar to wrap the raw event back into a CLoudEvent? What happens if this isn't set, will the app get a raw event without a Cloud Event?

Author: If this is not set, Dapr runtime will fail to parse the message and it will not work.

{{< tabs "Python SDK" "PHP SDK" >}}

{{% codetab %}}

```python
import flask
from flask import request, jsonify
from flask_cors import CORS
import json
import sys

app = flask.Flask(__name__)
CORS(app)

@app.route('/dapr/subscribe', methods=['GET'])
def subscribe():
    subscriptions = [{'pubsubname': 'pubsub',
                      'topic': 'deathStarStatus',

>                       'route': 'dsstatus',

Reviewer: Please add a dsstatus route methid in the code to show the use of the Dapr SDK. Currently there's no SDK in this example

Author: Done.

---
type: docs
title: "Pub/Sub without CloudEvents"
linkTitle: "Pub/Sub without CloudEvents"
weight: 7000
description: "Use Pub/Sub without CloudEvents." 
---

## Introduction

Dapr uses CloudEvents to provide additional context to the event payload, enabling features like:
* Tracing
* Deduplication by message Id
* Content-type for proper deserialization of event's data

For more information about CloudEvent, please check the [official specification](https://github.com/cloudevents/spec).

When adding Dapr to your application, some services may still need to communicate via raw pub/sub messages not encapsulated in CloudEvents. This may be for compatibility reasons, or because some apps are not using Dapr yet. Dapr enables apps to publish and subscribe to raw events that are not wrapped in a CloudEvent.

> 

Reviewer: ```suggestion

<img src="/images/pubsub_raw.png" alt="Diagram showing how to publish with Dapr when subscriber does not use Dapr or CloudEvent" width=1000>

```

Reviewer: One single image is better at the top vs. two images is subsections.

Author: Actually, these are two different scenarios. Raw pubsub is not supposed to be used when both applications use Dapr.

{{% alert title="Warning" color="warning" %}}
Not using CloudEvents disables support for tracing, event deduplication per messageId, content-type metadata, and any other features built using the CloudEvent schema.
{{% /alert %}}

## Publishing raw messages

> 

Reviewer: ```suggestion

Dapr apps are able to publish raw events to pub/sub topics without CloudEvent encapsulation, for compatibility with non-Dapr apps.

```

> <img src="/images/pubsub_publish_raw.png" alt="Diagram showing how to publish with Dapr when subscriber does not use Dapr or CloudEvent" width=1000>

Reviewer: ![publisher](https://user-images.githubusercontent.com/54363786/119699207-c179d380-be06-11eb-8111-7c90832576ad.png)

Author: Thanks for the nice image. Added it.

> <img src="/images/pubsub_publish_raw.png" alt="Diagram showing how to publish with Dapr when subscriber does not use Dapr or CloudEvent" width=1000>
> 

Reviewer: ```suggestion
```

{{< tabs curl "Python SDK" "PHP SDK">}}

{{% codetab %}}
```bash
curl -X "POST" http://localhost:3500/v1.0/publish/pubsub/TOPIC_A?metadata.rawPayload=true -H "Content-Type: application/json" -d '{"order-number": "345"}'
```
{{% /codetab %}}

{{% codetab %}}
```python
from dapr.clients import DaprClient

with DaprClient() as d:
    req_data = {
        'order-number': '345'
    }
    # Create a typed message with content type and body
    resp = d.publish_event(
        pubsub_name='pubsub',
        topic='TOPIC_A',
        data=json.dumps(req_data),
        metadata=(
                     ('rawPayload', 'true')
                 )
    )
    # Print the request
    print(req_data, flush=True)
```
{{% /codetab %}}

{{% codetab %}}

```php
<?php

require_once __DIR__.'/vendor/autoload.php';

$app = \Dapr\App::create();
$app->run(function(\DI\FactoryInterface $factory) {
    $publisher = $factory->make(\Dapr\PubSub\Publish::class, ['pubsub' => 'pubsub']);
    $publisher->topic('TOPIC_A')->publish('data', ['rawPayload' => 'true']);
});
```

{{% /codetab %}}

{{< /tabs >}}

> ## Subscribing to raw messages

Reviewer: Is there a declarative option?

Author: I just figured it out that our Subscription CRD lacks the metadata attribute:

https://github.com/dapr/dapr/issues/3225

Author: BTW, I've added a note in this doc mentioning that this feature cannot be used in the declarative way because of that.

> 

Reviewer: ```suggestion

Dapr apps are also able to subscribe to raw events coming from existing pub/sub topics that do not use CloudEvent encapsulation.

```

> <img src="/images/pubsub_subscribe_raw.png" alt="Diagram showing how to subscribe with Dapr when publisher does not use Dapr or CloudEvent" width=1000>

Reviewer: ![subscriber](https://user-images.githubusercontent.com/54363786/119699658-3ea54880-be07-11eb-8c29-6034a9a2d74c.png)

Author: Thanks for the nice image. Added it.

> <img src="/images/pubsub_subscribe_raw.png" alt="Diagram showing how to subscribe with Dapr when publisher does not use Dapr or CloudEvent" width=1000>

Reviewer: ```suggestion
```

When subscribing programmatically, add the additional metadata entry for `rawPayload` so the Dapr sidecar automatically wraps the payloads into a CloudEvent that is compatible with current Dapr SDKs.

> {{< tabs "Python SDK" "PHP SDK" >}}

Reviewer: ```suggestion
{{< tabs "Python" "PHP SDK" >}}
```

Doesn't use SDK

```
{{% /codetab %}}

{{% codetab %}}
```python
from dapr.clients import DaprClient

with DaprClient() as d:
    req_data = {
        'order-number': '345'
    }
    # Create a typed message with content type and body
    resp = d.publish_event(
        pubsub_name='pubsub',
        topic='TOPIC_A',
        data=json.dumps(req_data),
        metadata=(
                     ('rawPayload', 'true')
                 )
    )
    # Print the request
    print(req_data, flush=True)
```
{{% /codetab %}}

{{% codetab %}}

```php
<?php

require_once __DIR__.'/vendor/autoload.php';

$app = \Dapr\App::create();
$app->run(function(\DI\FactoryInterface $factory) {
    $publisher = $factory->make(\Dapr\PubSub\Publish::class, ['pubsub' => 'pubsub']);
    $publisher->topic('TOPIC_A')->publish('data', ['rawPayload' => 'true']);
});
```

{{% /codetab %}}

{{< /tabs >}}

## Subscribing to raw messages

Dapr apps are also able to subscribe to raw events coming from existing pub/sub topics that do not use CloudEvent encapsulation.

<img src="/images/pubsub_subscribe_raw.png" alt="Diagram showing how to subscribe with Dapr when publisher does not use Dapr or CloudEvent" width=1000>

> When subscribing programmatically, add the additional metadata entry for `rawPayload` so the Dapr sidecar automatically wraps the payloads into a CloudEvent that is compatible with current Dapr SDKs.

Reviewer: ```suggestion

### Programmatically subscribe to raw events 

When subscribing programmatically, add the additional metadata entry for `rawPayload` so the Dapr sidecar automatically wraps the payloads into a CloudEvent that is compatible with current Dapr SDKs.
```

def subscribe():
    subscriptions = [{'pubsubname': 'pubsub',
                      'topic': 'deathStarStatus',
                      'route': 'dsstatus',
                      'metadata': {
                          'rawPayload': 'true',
                      } }]
    return jsonify(subscriptions)

@app.route('/dsstatus', methods=['POST'])
def ds_subscriber():
    print(request.json, flush=True)
    return json.dumps({'success':True}), 200, {'ContentType':'application/json'}

app.run()
```

{{% /codetab %}}
{{% codetab %}}

```php
<?php

require_once __DIR__.'/vendor/autoload.php';

$app = \Dapr\App::create(configure: fn(\DI\ContainerBuilder $builder) => $builder->addDefinitions(['dapr.subscriptions' => [
    new \Dapr\PubSub\Subscription(pubsubname: 'pubsub', topic: 'deathStarStatus', route: '/dsstatus', metadata: [ 'rawPayload' => 'true'] ),
]]));

$app->post('/dsstatus', function(
    #[\Dapr\Attributes\FromBody]
    \Dapr\PubSub\CloudEvent $cloudEvent,
    \Psr\Log\LoggerInterface $logger
    ) {
        $logger->alert('Received event: {event}', ['event' => $cloudEvent]);
        return ['status' => 'SUCCESS'];
    }
);

$app->start();
```
{{% /codetab %}}

{{< /tabs >}}

> You cannot use this feature when subscribing to a topic using the Subscription Custom Resources Definition (CRD) since it does not contain the metadata attribute.

Reviewer: ```suggestion

## Declaratively subscribe to raw events

Subscription Custom Resources Definitions (CRDs) do not currently contain metadata attributes ([issue #3225](https://github.com/dapr/dapr/issues/3225)). At this time subscribing to raw events can only be done through programmatic subscriptions.
```
